README.md
<p align="center">
  <img width="144px" src="/public/logo.svg" />
</p>
<h1 align="center">vite-standard-template</h1>
<p align="center">一款简洁，符合通用标准的，基于Vite构建的Vue3项目模板</p>
<p align="center">
  <a href="https://npmjs.com/package/vite"><img src="/public/version/vite.svg" alt="vite version"></a>
    <a href="https://npmjs.com/package/vue"><img src="/public/version/vue.svg" alt="vue version"></a>
  <a href="https://nodejs.org/en/about/releases/"><img src="/public/version/node.svg" alt="node compatibility"></a>
</p>

## 一 项目简介

对于多人协作开发来说，项目开发规范标准，非常重要，不仅能够统一标准，规范代码书写风格，还能便于后期维护和运营。

一直以来，每次采用`vite`搭建`Vue3`项目时，都苦于配置`Eslint`代码校验规范，网上的水贴层出不穷，你抄我的，我粘贴你的，没有几个是有用的，不仅容易误导他人，还浪费大量的时间排查问题。最近花了一些时间，查阅了很多资料，自己也通过反复验证，决定自己搭建一个简单的模板，集成好日常代码开发规范和提交规范，把常用
的`Vite`插件都配置好，在以后创建项目时直接使用，避免重复造轮子，同时也供他人参考。

这是一个基于Vite4.X + Vue3.X + TypeScript + Naive UI + Pinia + VueRouter + Unocss + Alova + Eslint + Prettier + husky + lint-staged + commitlint + commitizen + cz-customizable+ conventional-changelog构建的标准的vue项目模板。整个项目包依赖采用`pnpm`进行依赖管理，`Node`版本为16.17.0, 同时也集成了项目中常用的插件，包含组件自动导入API

## 二 项目创建

1 采用`pnpm`创建项目

```shell
pnpm create vite
```

2 依赖安装

```shell
pnpm install
```

3 项目运行

```shell
pnpm dev
```

## 三 设置依赖
项目创建完后，我们开始增加一些日常开发过程中经常使用的`vite`插件和项目工具包，方便我们能够快速开发项目。
### 1 Vite插件
#### 1 unplugin-auto-import
> 官网：https://github.com/antfu/unplugin-auto-import

`unplugin-auto-import`是为 `Vite、Webpack、Rollup` 和 `esbuild` 按需自动导入`API`。例如：`ref,reactive`等API无需额外导入，就可以全局使用。
##### 1.1 依赖安装
```shell
pnpm install unplugin-auto-import -D
```

##### 1.2 插件配置
在`vite.config.ts`中进行插件配置，如下：
```ts
export default defineConfig(){
    plugins:[
      ...
       //自动导入Composition API,https://github.com/antfu/unplugin-auto-import
      AutoImport({
        dts: "src/types/auto-import.d.ts",
        imports: [
          "vue"
        ],
      }),
      ...
    ]
}
```
上面已经配置了`vue`框架自动导入API

#### 2 unplugin-vue-components
>官网：https://github.com/antfu/unplugin-vue-components

`unplugin-vue-components`是一款组件自动导入`Vite`插件，可以自定义需要自动导入的组件目录，无需使用时手动导入。

##### 2.1 依赖安装
```shell
pnpm install unplugin-vue-components -D
```
##### 2.2 插件配置
在`vite.config.ts`中进行插件配置，如下：
```ts
export default defineConfig(){
    plugins:[
      ...
        //自动导入组件，https://github.com/antfu/unplugin-vue-components
        Components({
          dts: "src/types/components.d.ts",
          dirs: ['src/components'],
          resolvers: [],
        }),
      ...
    ]
}
```
#### 3 unplugin-vue-setup-extend-plus
> 官网：https://github.com/chenxch/unplugin-vue-setup-extend-plus

`Vue3`组件自定义命名插件，可以在`<script setup lang=ts name="Good"></script>`标签中，通过设置name属性为组件命名

##### 3.1 依赖安装
```shell
pnpm install unplugin-vue-setup-extend-plus -D
```
##### 3.2 插件配置
在`vite.config.ts`中进行插件配置，如下：
```ts
export default defineConfig(){
    plugins:[
      ...
      //官网：https://github.com/chenxch/unplugin-vue-setup-extend-plus
      vueSetupExtend({
        //禁止组件属性自动导出
        enableAutoExpose: false,
      }),
      ...
    ]
}
```
配置完毕后，我们可以为组件进行命名。例如:`SvgIcon`组件：
```vue
<script setup lang="ts" name="SvgIcon">
</script>
```
#### 4 vite-plugin-html
> 官网地址：https://github.com/vbenjs/vite-plugin-html/blob/main/README.zh_CN.md

`vite-plugin-html`插件可以在`html`页面中使用`ejs`语法，动态注入数据。

##### 4.1 依赖安装
```shell
pnpm install vite-plugin-html -D
```
##### 4.2 插件配置
1 在`vite.config.ts`中进行插件配置，如下：
```ts
export default defineConfig(){
    plugins:[
      ...
      //在html中创建ejs标签，官网地址：https://github.com/vbenjs/vite-plugin-html/blob/main/README.zh_CN.md
      createHtmlPlugin({
        // 是否压缩 html
        minify: true,
        /**
         * 需要注入 index.html ejs 模版的数据
         */
        inject: {
          data: {
            title: env.VITE_SYSTEM_TITLE,
            description: env.VITE_SYSTEM_DESC
          }
        }
      }),
      ...
    ]
}
```
`inject`中的data就是要注入的变量参数，`env`为环境变量参数。可以通过一下代码获取到：
```ts
import { loadEnv } from 'vite'
const env = loadEnv(mode, process.cwd())
```
2 修改`index.html`文件，将`vite-plugin-html`插件注入的数据，通过`ejs`语法写入`index.html`,如下:
```html
<!doctype html>
<html lang="zh-CN">
    <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <meta name="description" content="<%=description%>" />
        <link rel="icon" type="image/svg+xml" href="/logo.svg" />
        <title><%= title %></title>
    </head>

    <body>
        <div id="app">
            <div id="loading"></div>
        </div>
        <script type="module" src="/src/main.ts"></script>
    </body>
</html>
```
#### 5 vite-plugin-svg-icons
> 官网:https://github.com/vbenjs/vite-plugin-svg-icons/blob/main/README.zh_CN.md

`vite-plugin-svg-icons`是一款用于生成`svg` 雪碧图的插件，能够将本地指定文件目录下的`Svg`生成一张雪碧图，在项目运行时就生成所有图标,只需操作一次`dom`,内置缓存,仅当文件被修改时才会重新生成。通过`Svg`名称便可以加载对应的`Svg`图标。

##### 5.1 依赖安装
```shell
pnpm install vite-plugin-svg-icons -D
```
##### 5.2 插件配置
1 在`vite.config.ts`中进行插件配置，如下：
```ts
import { createSvgIconsPlugin } from 'vite-plugin-svg-icons'
import path from 'path'
export default defineConfig(){
    plugins:[
      ...
      createSvgIconsPlugin({
       //// 指定需要缓存的svg图标文件夹
       iconDirs: [path.resolve(process.cwd(), 'src/assets/svg')],
       // 指定symbolId格式
       symbolId: 'icon-local-[dir]-[name]`,
       // 自定义插入位置,@default: body-last
       inject: 'body-last',
       //自定义domId,默认：__svg__icons__dom__
       customDomId: '__SVG_ICON_LOCAL__'
      })
      ...
    ]
}
```
2 在`main.js`中增加以下代码
```js
import 'virtual:svg-icons-register'
```
3 将需要的`svg`图标放入与`iconDirs`设置的路径中，项目中为`src/assets/svg`
```svg
<svg aria-hidden="true" style="width: 14px; height: 14px">
	<use :href="`#icon-local-${menu.icon}`" />
</svg>
```
`menu.icon`是路径里面的`svg`图片名称。这个是简单用法，项目中已封装成了组件`SvgIcon`,请前往自行查看。

4 如果使用`Typescript`,你可以在`tsconfig.json`内添加:
```json
// tsconfig.json
{
  "compilerOptions": {
    "types": ["vite-plugin-svg-icons/client"]
  }
}
```
#### 6 @iconify/vue
> 官网：https://iconify.design/docs/icon-components/vue/

`iconify`是功能最丰富的图标框架。可以与任何图标库一起使用的统一图标框架。开箱即用的功能包括80多个图标集和超过70,000个图标。官方为了便于使用`iconify`网站上的图标，提供了`@iconify/vue`组件，供大家使用`SVG framework`，支持在线和离线2种方式使用。离线方式需要下载对应图标集合`json`数据，然后先从本地资源中加载，如果没有找到，通过API从线上下载资源，并进行浏览器缓存。

`@iconify/vue`是一个功能非常强大的组件，支持图标名称动态渲染和静态渲染，正好弥补`vite-plugin-svg-icons`功能缺陷。
##### 6.1 依赖安装
```shell
pnpm install @iconify/vue -D
```
##### 6.2 使用例子
```ts
import { Icon } from '@iconify/vue';
<Icon icon="mdi-light:home" />
```
##### 6.3 自定义组件
虽然`@iconify/vue`也支持本地`svg`,但逐个配置非常麻烦，所以结合`vite-plugin-svg-icons`和`@iconify/vue`,我们创建一个自定义组件，使其不仅支持本地静态动态`Svg`渲染，还支持显示静态动态`Svg`渲染。下面是自定义组件代码：
```vue
<template>
    <template v-if="localIcon">
        <svg aria-hidden="true" width="1em" height="1em" v-bind="bindAttrs">
            <use :xlink:href="symbolId" fill="currentColor" />
        </svg>
    </template>
    <template v-else>
        <Icon v-if="icon" :icon="icon" v-bind="bindAttrs" />
    </template>
</template>

<script setup lang="ts" name="SvgIcon">
import { Icon } from '@iconify/vue'
// eslint-disable-next-line vue/no-setup-props-destructure
const { icon, localIcon } = defineProps<{
    /** iconify线上图标名称 */
    icon?: string
    /** 本地svg的文件名称 */
    localIcon?: string
}>()
// 获取组件传递的属性
const attrs = useAttrs()
// 计算绑定属性
const bindAttrs = computed<{ class: string; style: string }>(() => ({
    class: (attrs.class as string) ?? 'w-24px h-24px',
    style: attrs.style as string
}))
// 计算本地svg动态的symbolId
const symbolId = computed(() => {
    const icon = localIcon ?? 'no-icon'
    return `#icon-local-${icon}`
})
</script>
```
此时，便可以灵活的渲染本地和线上`iconify`网站上的`Svg`图标，使用例子如下：
```vue
本地Svg:<SvgIcon localIcon="logo"></SvgIcon>
iconify线上Svg:<SvgIcon icon="healthicons:fhir-logo" class="w-24px h-24px"></SvgIcon>
```

#### 7 rollup-plugin-visualizer
> 官网：https://github.com/btd/rollup-plugin-visualizer

`rollup-plugin-visualizer`是一个用于Rollup构建工具的插件，一款用于项目性能优化，打包体积分析，能够生成可视化的构建报告，帮助开发者更好地了解构建过程中的文件大小、依赖关系等信息的插件。

##### 7.1 依赖安装
```shell
pnpm install rollup-plugin-visualizer -D
```

##### 7.2 插件配置

1 在`vite.config.ts`中进行插件配置，如下：
```ts
// 引入rollup-plugin-visualizer模块
import { visualizer } from "rollup-plugin-visualizer";
export default defineConfig(){
    plugins:[
      ...
      visualizer({
        //注意这里要设置为true，打包时会自动打开分析页面。
        open:true,  
        //分析图生成的文件名
        filename: "stats.html", 
        //收集gzip大小并将其显示
        gzipSize: true, // 
        //收集 brotli 大小并将其显示
        brotliSize: true, 
      })
      ...
    ]
}
```
2 运行运行命令打包，生成分析图

输入`pnpm build`打包项目，等待片刻，生成分析视图，视图会自动跳出来，并保存在项目根目录下，文件名就是刚刚参数filename的名字（stats.html）

![](https://cdn.staticaly.com/gh/AnyFork/blog-images/main/markdown/202307211635849.png)

视图分析中方块越大，表示该文件占用的空间越大，对于网络带宽和访问速度的要求就越高。如果一个网站中包含大量的大文件，那么用户在访问该网站时需要下载更多的数据，这会导致网站加载速度变慢，用户体验变差。

#### 8 vite-plugin-compression
> 官网：https://github.com/vbenjs/vite-plugin-compression

`gzip`压缩：当前端资源过大时，服务器请求资源会比较慢。前端可以将资源通过`Gzip`压缩使文件体积减少大概60%左右，压缩后的文件，通过后端简单处理，浏览器可以将其正常解析出来。如果浏览器的请求头中包含`content-encoding: gzip`，即证明浏览器支持该属性。

`vite`中使用`vite-plugin-compression`插件可以很便捷的对代码进行`gzip`压缩，减少代码体积，加快浏览器访问速度。压缩的代码放到服务器后，需要后端配置一些东西，浏览器才可以解析。比如可以配置nginx.

##### 8.1 依赖安装
```shell
pnpm install vite-plugin-compression -D
```
##### 8.2 插件配置
在`vite.config.ts`中进行插件配置，如下：
```ts
// 引入vite-plugin-compression模块
import viteCompression from 'vite-plugin-compression';
export default defineConfig(){
    plugins:[
      ...
      viteCompression()
      ...
    ]
}
```
#### 9 完整的vite.config.ts
下面展示整个项目`vite.config.ts`完整的配置，代码如下
```ts
//vite.config.ts
import { type ConfigEnv, defineConfig, loadEnv } from 'vite'
import vue from '@vitejs/plugin-vue'
// 该包是用于配置vite运行的时候自动检测eslint规范,不符合规范，启动时不会报错，页面刷新时会报错，https://github.com/gxmari007/vite-plugin-eslint
import eslint from 'vite-plugin-eslint'
import { visualizer } from 'rollup-plugin-visualizer'
import ViteCompression from 'vite-plugin-compression'
import Components from 'unplugin-vue-components/vite'
import { NaiveUiResolver } from 'unplugin-vue-components/resolvers'
import { createSvgIconsPlugin } from 'vite-plugin-svg-icons'
import AutoImport from 'unplugin-auto-import/vite'
import UnoCSS from 'unocss/vite'
import { createHtmlPlugin } from 'vite-plugin-html'
import path from 'path'

// https://vitejs.dev/config/
export default defineConfig(({ mode }: ConfigEnv) => {
    const env = loadEnv(mode, process.cwd())
    return {
        // 配置插件
        plugins: [
            vue(),
            eslint(),
            /** 打包分析插件，官网：https://github.com/btd/rollup-plugin-visualizer */
            visualizer({
                // 注意这里要设置为true，打包时会自动打开分析页面。
                open: true,
                // 分析图生成的文件名
                filename: 'stats.html',
                // 收集gzip大小并将其显示
                gzipSize: true, //
                // 收集 brotli 大小并将其显示
                brotliSize: true
            }),
            /** 打包压缩插件 官网：https://github.com/vbenjs/vite-plugin-compression */
            ViteCompression({ algorithm: 'gzip' }),
            // 自动导入Composition API,https://github.com/antfu/unplugin-auto-import
            AutoImport({
                // dts生成路径
                dts: 'src/types/auto-import.d.ts',
                // 自动本地导入文件目录路径
                dirs: ['src/store/modules'],
                // 设置第三方自动导入的包名
                imports: [
                    'vue',
                    'vue-router',
                    'pinia',
                    '@vueuse/core',
                    {
                        'naive-ui': ['useDialog', 'useMessage', 'useNotification', 'useLoadingBar'],
                        alova: ['useRequest', 'createAlova'],
                        '@/service/alova': ['alova']
                    }
                ]
            }),
            // 自动导入组件，https://github.com/antfu/unplugin-vue-components
            Components({
                dts: 'src/types/components.d.ts',
                dirs: ['src/components'],
                resolvers: [NaiveUiResolver()]
            }),
            // 本地svg动态加载插件 官网地址：https://github.com/vbenjs/vite-plugin-svg-icons
            createSvgIconsPlugin({
                iconDirs: [path.resolve(process.cwd(), 'src/assets/svg')],
                symbolId: 'icon-local-[dir]-[name]',
                inject: 'body-last',
                customDomId: '__SVG_ICON_LOCAL__'
            }),
            // 官网地址:https://unocss.dev/integrations/vite
            UnoCSS(),
            // 在html中创建ejs标签，官网地址：https://github.com/vbenjs/vite-plugin-html/blob/main/README.zh_CN.md
            createHtmlPlugin({
                // 是否压缩 html
                minify: true,
                /**
                 * 需要注入 index.html ejs 模版的数据
                 */
                inject: {
                    data: {
                        title: env.VITE_SYSTEM_TITLE,
                        description: env.VITE_SYSTEM_DESC,
                        keywords: env.VITE_SYSTEM_KEYWORDS
                    }
                }
            })
        ],
        resolve: {
            // 配置别名
            alias: [{ find: '@', replacement: path.resolve(__dirname, 'src') }]
        },
        // 本地运行配置，及跨域反向代理配置
        server: {
            // 服务器端口
            port: 9527,
            // 默认启用并允许任何源
            cors: true,
            // 在服务器启动时自动在浏览器中打开应用程序
            open: true,
            // 反向代理配置，注意rewrite写法，开始没看文档在这里踩了坑
            proxy: {
                // 本地开发环境通过代理实现跨域，生产环境使用nginx转发
                '/api': {
                    // 通过代理接口访问实际地址。这里是实际访问的地址。vue会通过代理服务器来代理请求
                    target: env.VITE_BASIC_API_URL,
                    changeOrigin: true,
                    // 允许websocket代理
                    ws: true,
                    // 将api替换为空
                    rewrite: (path) => path.replace(/^\/api/, '/api')
                }
            }
        }
    }
})
```

### 2 Naive-UI框架
> 官网地址：https://www.naiveui.com/zh-CN/os-theme/docs/installation

`naive-ui`全量使用`TypeScript`编写, 无样式文件，组件按需加载，是一个使用起来非常`清爽`的`Vue 3`组件库, 组件比较完整，主题可调，完全兼容现在主流的浏览器。

#### 1 安装依赖
```shell
pnpm install naive-ui -D
```
#### 2 按需引入
可以使用`unplugin-auto-import` 插件来自动导入`API`,使用`unplugin-vue-components`插件来按需自动加载组件，插件会自动解析模板中的使用到的组件，并导入组件。

1 在`vite.config.ts`中进行配置，代码如下：
```js
import AutoImport from 'unplugin-auto-import/vite'
import Components from 'unplugin-vue-components/vite'
import { NaiveUiResolver } from 'unplugin-vue-components/resolvers'
// https://vitejs.dev/config/
export default defineConfig({
  plugins: [
    vue(),
    AutoImport({
      imports: [
        'vue',
        {
          'naive-ui': [
            'useDialog',
            'useMessage',
            'useNotification',
            'useLoadingBar'
          ]
        }
      ]
    }),
    Components({
      resolvers: [NaiveUiResolver()]
    })
  ]
})
```
2 如果你在使用`Volar`，那么可以在`tsconfig.json`中配置`compilerOptions.types`来指定全局组件类型
```js
// tsconfig.json
{
  "compilerOptions": {
    // ...
    "types": ["naive-ui/volar"]
  }
}
```
至此`naive-ui`按需导入成功，此时便可以使用相关组件进行开发，详情参考：[naive-ui官网](https://www.naiveui.com/zh-CN/os-theme/docs/installation)


### 3 Unocss引擎
> 官方网站：https://unocss.dev/

`UnoCSS` - 一个具有高性能且极具灵活性的即时原子化`CSS`引擎，而非一款框架，因为它并未提供核心工具类，所有功能可以通过预设和内联配置提供，目前内置的`presetUno`预设涵盖了[Tailwindcss](https://www.tailwindcss.cn/)和[Windicss](https://cn.windicss.org/)的大部分功能，熟悉上面2个原子css框架写法的人，上手很容易，`UnoCSS`的主要目标是直观性和可定制性。
#### 3.1 依赖安装
```shell
pnpm install unocss  -D
```
#### 3.2 引擎配置
在`vite.config.ts`增加如下配置，相关配置参考：[unocss](https://unocss.dev/integrations/vite)
```ts
// vite.config.ts
import UnoCSS from 'unocss/vite'
import { defineConfig } from 'vite'
export default defineConfig({
  plugins: [
    ...
    UnoCSS(),
    ...
  ],
})
```
#### 3.2 创建uno.config.ts
在项目根目录下面创建`uno.config.ts`,增加如下配置：
```ts
//uno.config.ts
import { defineConfig, presetUno, presetAttributify } from 'unocss'
export default defineConfig({
    //预设配置参考：https://unocss.dev/presets
    presets: [
        //设置默认预设，当自定义其他预设后，默认预设需要额外添加
        presetUno({ dark: "class" }),
        //设置归因预设(Attributify preset),可以使用bg=red等语法
        presetAttributify()
    ],
    //设置shortcuts,只能使用预设的和自定义的规则
    shortcuts: {
        'wh-full': 'w-full h-full',
        'flex-row-center': 'flex justify-center items-center',
        'flex-row-between': 'flex justify-between items-center',
        'flex-row-evenly': 'flex justify-evenly items-center',
        'flex-row-warp': 'flex flex-wrap',
        'flex-row-end': 'flex justify-end items-center',
        'flex-col-center': 'flex flex-col justify-center items-center',
        'flex-x-center': 'flex justify-center',
        'flex-y-center': 'flex items-center',
        'i-flex-center': 'inline-flex justify-center items-center',
        'i-flex-x-center': 'inline-flex justify-center',
        'i-flex-y-center': 'inline-flex items-center'
    },
    //自定义规则
    rules: [],
    //主题配置
    theme: {
        //继承boxShadow
        boxShadow: {
            box: '0 1px 8px 0 rgba(255, 0, 0, 0.1)',
            item: "0 1px 8px 0 rgba(0, 0, 0, 0.1)"
        },
        colors: {
            primary: 'rgb(var(--primary-color))',
            primary_hover: 'var(--primary-color-hover)',
            primary_pressed: 'var(--primary-color-pressed)',
            primary_active: 'var(--primary-color-active)',
            info: 'var(--info-color)',
            info_hover: 'var(--info-color-hover)',
            info_pressed: 'var(--info-color-pressed)',
            info_active: 'var(--info-color-active)',
            success: 'var(--success-color)',
            success_hover: 'var(--success-color-hover)',
            success_pressed: 'var(--success-color-pressed)',
            success_active: 'var(--success-color-active)',
            warning: 'var(--warning-color)',
            warning_hover: 'var(--warning-color-hover)',
            warning_pressed: 'var(--warning-color-pressed)',
            warning_active: 'var(--warning-color-active)',
            error: 'var(--error-color)',
            error_hover: 'var(--error-color-hover)',
            error_pressed: 'var(--error-color-pressed)',
            error_active: 'var(--error-color-active)'
        }
    }
})
```
上面是我项目中经常使用到的配置。

#### 3.3 配置virtual:uno.css
在`main.ts`中增加如下代码：
```ts
// main.ts
import 'virtual:uno.css'
```
#### 3.4 VScode插件安装
`VScode`插件市场有`Unocss`插件，安装以后鼠标放上去可以查看对应`class`的`css`属性和值信息，[点我安装插件](https://marketplace.visualstudio.com/items?itemName=antfu.unocss)

### 4 Pinia状态管理
> 官网：https://pinia.vuejs.org/

`Pinia`是`Vue`的存储库，它允许您跨组件/页面共享状态，与`vuex`功能类似，不仅兼容`Option API`写法也兼容`Composition API`写法，同时也只是插件扩展。

#### 4.1 依赖安装
```shell
pnpm install pinia
```
#### 4.2 配置Pinia实例
1、在`vite.config.ts`文件的`AutoImport`自动导入插件中，增加`pinia`API自动导入。
```ts
export default defineConfig(({ mode }: ConfigEnv) => {
  return {
    plugins: [
      ...
      //自动导入Composition API,https://github.com/antfu/unplugin-auto-import
      AutoImport({
        dts: "src/types/auto-import.d.ts",
        imports: [
          ...
          + "pinia",
          ...
        ],
      })
      ...
    ]
  }
})
```
配置了`pinia`API自动导入后，`defineStore`和`createPinia`无需导入，便可以直接使用。

2 在项目`src`目录下创建`store`目录，创建`index.ts`文件，配置`pinia`实例。
```ts
//src/store/index.ts
import type { App } from 'vue';
/**
 * 安装vue状态管理插件pinia
 * @param app 
 */
export function setupStore(app: App) {
    const pinia = createPinia();
    //挂载pinia实例到app
    app.use(pinia);
}
```
3 在`main.ts`中注册`pinia`实例插件。
```ts
//main.ts
import { createApp } from 'vue'
import './style.css'
import 'virtual:uno.css'
+ import { setupStore } from './store';
import App from './App.vue'
const setupApp = async () => {
    //创建vue实例
    const app = createApp(App)
    //创建pinia
    + setupStore(app);
    //挂载app
    app.mount('#app');
}
setupApp()
```
到此为止，`pinia`实例创建好了，并挂载在了`vue`实例上。

4 创建`src/store/modules`目录，将状态管理文件按照模块放在`src/store/modules`目录中。建议使用`Composition API`写法进行开发。下面是个简单例子:
```ts
//src/store/modules/demo.ts
export const useCounterStore = defineStore('counter', () => {
    const count = ref(0)
    const increment = (): void => {
        count.value++
    }
    return { count, increment }
})
```
请注意，`store`是一个用`reactive`包裹的对象，这意味着不需要在`getter`之后写`.value`，但是，就像`setup` 中的`props` 一样，我们不能对其进行解构：
```ts
export default defineComponent({
  setup() {
    const store = useStore()
    // ❌ 这不起作用，因为它会破坏响应式
    // 这和从 props 解构是一样的
    const { name, doubleCount } = store

    name // "eduardo"
    doubleCount // 2

    return {
      // 一直会是 "eduardo"
      name,
      // 一直会是 2
      doubleCount,
      // 这将是响应式的
      doubleValue: computed(() => store.doubleCount),
      }
  },
})
```
为了从`Store`中提取属性同时保持其响应式，您需要使用`storeToRefs()`。 它将为任何响应式属性创建 `refs`。 当您仅使用`store`中的状态但不调用任何操作时，这很有用：
```ts
import { storeToRefs } from 'pinia'
export default defineComponent({
  setup() {
    const store = useStore()
    // `name` 和 `doubleCount` 是响应式引用
    // 这也会为插件添加的属性创建引用
    // 但跳过任何 action 或 非响应式（不是 ref/reactive）的属性
    const { name, doubleCount } = storeToRefs(store)

    return {
      name,
      doubleCount
    }
  },
})
```

5 将模块目录`modules`下面的文导出到`src/store/modules/index.ts`文件中，这样我们以后使用的时候，可以直接从`@/store/modules`直接导入
```ts
//src/store/modules/index.ts
export * from './demo'
//src/store/index.ts
export * from './modules'
```
5 在需要调用的文件中，通过命令导入`import { useCounterStore } from '@/store/modules'`,使用方法如下：
```ts
import { useCounterStore } from '@/store/modules'
const counter = useCounterStore()
const { count } = storeToRefs(counter)
```
注意，不能直接对`useCounterStore()`进行结构，会影响响应性，需要用`storeToRefs`进行包装，然后就可以在模板中使用`count`和`increment`了。

6 `AutoImport`设置自动导入

虽然经过上面的配置，我们可以通过`import { useCounterStore } from '@/store/modules'`在需要的文件中进行导入使用，但每次都要写就显的麻烦，可以通过功能强大的`AutoImport`插件帮我们自动导入，以后就可以直接使用`useCounterStore`,就不用导入了。

在`vite.config.ts`文件的`AutoImport`自动导入插件中，增加`src/store/modules`目录API的自动导入。
```ts
export default defineConfig(({ mode }: ConfigEnv) => {
  return {
    plugins: [
      ...
      //自动导入Composition API,https://github.com/antfu/unplugin-auto-import
      AutoImport({
        dts: "src/types/auto-import.d.ts",
        // 自动本地导入文件目录路径
        + dirs: ['src/store/modules'],
        imports: [
          ...
          "pinia",
          ...
        ],
      })
      ...
    ]
  }
})
```
到此，便可以愉快的玩耍了:smile: 
#### 4.3 状态持久化
`pinia`支持插件扩展，增强自身功能。类似`vuex`一样，`pinia`也有状态持久化插件`pinia-plugin-persistedstate`,帮助`pinia`完成状态持久功能，支持`localStorage`和`sessionStorage`2种持久化方式。
> 官网: https://prazdevs.github.io/pinia-plugin-persistedstate/guide/

##### 4.3.1 依赖安装
```shell
pnpm install pinia-plugin-persistedstate
```
##### 4.3.2 插件配置
1 在`src/store/`目录下面创建文件目录`plugins`, 将插件`pinia-plugin-persistedstate`配置代码放在`src/store/plugin/modules/persistedstate.ts`中。
```ts
// src/store/plugin/modules/persistedstate.ts
// 参考地址：https://prazdevs.github.io/pinia-plugin-persistedstate/guide/
import { createPersistedState } from 'pinia-plugin-persistedstate'
import { encrypto, decrypto } from '@/utils'
/**
 * pinia 全局持久化配置，会覆盖默认配置，但也会被单个store的persist配置覆盖
 */
export const createPersistedStatePlugins = createPersistedState({
    storage: sessionStorage,
    beforeRestore: (context) => {
        console.log(context)
        return context
    },
    afterRestore: (context) => {
        console.log(context)
        return context
    },
    // 设置序列化，生产加密，开发采用默认不加密
    serializer: {
        serialize: import.meta.env.PROD ? encrypto : JSON.stringify,
        deserialize: import.meta.env.PROD ? decrypto : JSON.parse
    }
})
```
`serializer`默认的序列化为`JSON.stringify`,反序列化为`JSON.parse`，开发期间为了便于观察数据，采用默认的序列化方式，生成环境采用`encrypto`和`decrypto`进行加密和解密，增加数据的安全性。更多具体配置，请查看[官方文档](https://prazdevs.github.io/pinia-plugin-persistedstate/guide/)

2 将配置好的插件安装在`pinia`实例上面，配置如下：
```ts
import type { App } from 'vue'
+ import { createPersistedStatePlugins } from './plugins'
/**
 * 安装vue状态管理插件pinia
 * @param app
 */

CHANGELOG.md
## <small>0.0.1 (2023-07-24)</small>

* :bug: fix: 修复tabWidth缩进规则 ([5aa8cd2](https://github.com/AnyFork/vite-standard-template/commit/5aa8cd2))
* :pencil2: docs: 更新文档 | 增加.eslintignore,.prettierignore ([d771f19](https://github.com/AnyFork/vite-standard-template/commit/d771f19))
* :sparkles: feat: 增加插件和Naive-UI ([6a7c864](https://github.com/AnyFork/vite-standard-template/commit/6a7c864))
* :sparkles: feat: 增加git代码提交规范 ([9e49d54](https://github.com/AnyFork/vite-standard-template/commit/9e49d54))
* :sparkles: feat: 增加pinia和路由 ([0a15e8c](https://github.com/AnyFork/vite-standard-template/commit/0a15e8c))
* :sparkles: feat(router): 添加简单路由 ([177f950](https://github.com/AnyFork/vite-standard-template/commit/177f950))
* ✨ feat: 增加Eslint和perttier配置 ([99484f6](https://github.com/AnyFork/vite-standard-template/commit/99484f6))
* 🎉 init: 项目初始化 ([3a026c9](https://github.com/AnyFork/vite-standard-template/commit/3a026c9))
* Initial commit ([dc86fcc](https://github.com/AnyFork/vite-standard-template/commit/dc86fcc))





uno.config.ts
// uno.config.ts
import { defineConfig, presetUno, presetAttributify } from 'unocss'
export default defineConfig({
    // 预设配置参考：https://unocss.dev/presets
    presets: [
        // 设置默认预设，当自定义其他预设后，默认预设需要额外添加
        presetUno({ dark: 'class' }),
        // 设置归因预设(Attributify preset),可以使用bg=red等语法
        presetAttributify()
    ],
    // 设置shortcuts,只能使用预设的和自定义的规则
    shortcuts: {
        'wh-full': 'w-full h-full',
        'flex-row-center': 'flex justify-center items-center',
        'flex-row-between': 'flex justify-between items-center',
        'flex-row-evenly': 'flex justify-evenly items-center',
        'flex-row-warp': 'flex flex-wrap',
        'flex-row-end': 'flex justify-end items-center',
        'flex-col-center': 'flex flex-col justify-center items-center',
        'flex-x-center': 'flex justify-center',
        'flex-y-center': 'flex items-center',
        'i-flex-center': 'inline-flex justify-center items-center',
        'i-flex-x-center': 'inline-flex justify-center',
        'i-flex-y-center': 'inline-flex items-center'
    },
    // 自定义规则
    rules: [],
    // 主题配置
    theme: {
        // 继承boxShadow
        boxShadow: {
            box: '0 1px 8px 0 rgba(255, 0, 0, 0.1)',
            item: '0 1px 8px 0 rgba(0, 0, 0, 0.1)'
        },
        colors: {
            primary: 'rgb(var(--primary-color))',
            primary_hover: 'var(--primary-color-hover)',
            primary_pressed: 'var(--primary-color-pressed)',
            primary_active: 'var(--primary-color-active)',
            info: 'var(--info-color)',
            info_hover: 'var(--info-color-hover)',
            info_pressed: 'var(--info-color-pressed)',
            info_active: 'var(--info-color-active)',
            success: 'var(--success-color)',
            success_hover: 'var(--success-color-hover)',
            success_pressed: 'var(--success-color-pressed)',
            success_active: 'var(--success-color-active)',
            warning: 'var(--warning-color)',
            warning_hover: 'var(--warning-color-hover)',
            warning_pressed: 'var(--warning-color-pressed)',
            warning_active: 'var(--warning-color-active)',
            error: 'var(--error-color)',
            error_hover: 'var(--error-color-hover)',
            error_pressed: 'var(--error-color-pressed)',
            error_active: 'var(--error-color-active)'
        }
    }
})
